server/datastore/s3/common_file_store.go

```diff
@@ -0,0 +1,134 @@
+//
+//	<bucket>/<prefix>/bootstrap-packages/<fileID>
+type commonFileStore struct {
+	*s3store
+	pathPrefix string
+	fileLabel  string // how to call the file in error messages
+}
+
+// Get retrieves the requested file from S3.
+// It is important that the caller closes the reader when done.
+func (s *commonFileStore) Get(ctx context.Context, fileID string) (io.ReadCloser, int64, error) {
+	key := s.keyForFile(fileID)
+
+	req, err := s.s3client.GetObject(&s3.GetObjectInput{Bucket: &s.bucket, Key: &key})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeNoSuchKey, s3.ErrCodeNoSuchBucket, "NotFound":
+				return nil, int64(0), installerNotFoundError{}
+			}
+		}
+		return nil, int64(0), ctxerr.Wrapf(ctx, err, "retrieving %s from S3 store", s.fileLabel)
+	}
+	return req.Body, *req.ContentLength, nil
+}
+
+// Put uploads a file to S3.
+func (s *commonFileStore) Put(ctx context.Context, fileID string, content io.ReadSeeker) error {
+	if fileID == "" {
+		return errors.New("S3 file identifier is empty")
+	}
+
+	key := s.keyForFile(fileID)
+	_, err := s.s3client.PutObject(&s3.PutObjectInput{
+		Bucket: &s.bucket,
+		Body:   content,
+		Key:    &key,
+	})
+	return err
+}
+
+// Exists checks if a file exists in the S3 bucket for the ID.
+func (s *commonFileStore) Exists(ctx context.Context, fileID string) (bool, error) {
+	key := s.keyForFile(fileID)
+
+	_, err := s.s3client.HeadObject(&s3.HeadObjectInput{Bucket: &s.bucket, Key: &key})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeNoSuchKey, s3.ErrCodeNoSuchBucket, "NotFound":
+				return false, nil
+			}
+		}
+		return false, ctxerr.Wrapf(ctx, err, "checking existence of %s in S3 store", s.fileLabel)
+	}
+	return true, nil
+}
+
+func (s *commonFileStore) Cleanup(ctx context.Context, usedFileIDs []string) (int, error) {
+	usedSet := make(map[string]struct{}, len(usedFileIDs))
+	for _, id := range usedFileIDs {
+		usedSet[id] = struct{}{}
+	}
+
+	// ListObjectsV2 defaults to a max of 1000 keys, which is sufficient for the
+	// cleanup task - if more files are present, the next run will get another
+	// 1000 and will periodically complete the cleanups.
+	//
+	// Iterating over all pages would potentially take a long time and would make
+	// it more likely that a conflict arises, where an unused file becomes used
+	// again. This approach makes it only two API requests between the read of
+	// used files and the deletions.
+	prefix := path.Join(s.prefix, s.pathPrefix)
+	page, err := s.s3client.ListObjectsV2(&s3.ListObjectsV2Input{
+		Bucket: &s.bucket,
+		Prefix: &prefix,
+	})
+	if err != nil {
+		return 0, ctxerr.Wrapf(ctx, err, "listing %s in S3 store", s.fileLabel)
+	}
+
+	// TODO(mna): there is an inherent risk that we could delete files that were
+	// added between the query to list used IDs and now. We could minimize that
+	// risk by checking that the S3 file is older than a few seconds.
```

Comment:
I think this is worth adding, I'll do it next week. It should make it practically impossible to delete S3 files that are in use (except in extreme cases where S3 API is super slow or the clock difference between S3 and our servers is significant, which I assume is very unlikely if Fleet is hosted in AWS itself).

Reply:
S3 has consistency guarantees:

> Amazon S3 provides strong read-after-write consistency for PUT and DELETE requests of objects in your Amazon S3 bucket in all AWS Regions. This behavior applies to both writes to new objects as well as PUT requests that overwrite existing objects and DELETE requests. In addition, read operations on Amazon S3 Select, Amazon S3 access controls lists (ACLs), Amazon S3 Object Tags, and object metadata (for example, the HEAD object) are strongly consistent.

https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html#ConsistencyModel

Not sure if it helps in this case, but figured I'd at least mention it.